core/src/test/scala/fs2/job/JobManagerSpec.scala

```diff
@@ -0,0 +1,163 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fs2
+package job
+
+import scala.Predef.String
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration._
+import scala.util.{Either, Left, Right}
+import scala.{Int, List, Unit}
+
+import cats.Eq
+import cats.effect.IO
+import cats.instances.string._
+import fs2.concurrent.SignallingRef
+import org.specs2.mutable._
+
+object JobManagerSpec extends Specification {
+  implicit val cs = IO.contextShift(ExecutionContext.global)
+  implicit val timer = IO.timer(ExecutionContext.global)
+
+  def await: Stream[IO, Unit] =
+    Stream.sleep(100.milliseconds)
+
+  def mkJobManager: Stream[IO, JobManager[IO, Int, String]] =
+    JobManager[IO, Int, String]()
+
+  "Job manager" should {
+    "submit a job" in {
+      val JobId = 1
+
+      val stream = await.as(Right(1)) ++ Stream(2, 3, 4).map(Right(_)).covary[IO]
+      val notification = Stream(Left("done")).covary[IO]
+
+      val job = Job[IO, Int, String, Int](JobId, stream ++ notification)
+
+      val (submitResult, ids, status) = (for {
+        mgr <- mkJobManager
+        submitResult <- Stream.eval(mgr.submit(job))
+        _ <- await
+        ids <- Stream.eval(mgr.jobIds)
+        status <- Stream.eval(mgr.status(JobId))
+      } yield (submitResult, ids, status)).compile.lastOrError.timeout(5.seconds).unsafeRunSync
+
+      submitResult must beTrue
+      ids must_== List(JobId)
+      status must beSome(Status.Running)
+    }
+
+    "execute a job to completion" in {
+      val JobId = 42
+
+      def jobStream(ref: SignallingRef[IO, String]): Stream[IO, Either[String, Int]] =
+        Stream.eval(ref.set("Started")).as(Right(1)) ++ await.as(Right(2)) ++ Stream.eval(ref.set("Finished")).as(Right(3))
+
+      val (refAfterStart, refAfterRun) = (for {
+        mgr <- mkJobManager
+        ref <- Stream.eval(SignallingRef[IO, String]("Not started"))
+        job = Job[IO, Int, String, Int](JobId, jobStream(ref))
+
+        submitStatus <- Stream.eval(mgr.submit(job))
+        _ <- latchGet(ref, "Started")
+        refAfterStart <- Stream.eval(ref.get)
+
+        _ <- latchGet(ref, "Finished")
+        refAfterRun <- Stream.eval(ref.get)
+      } yield (refAfterStart, refAfterRun)).compile.lastOrError.timeout(5.seconds).unsafeRunSync
+
+      refAfterStart mustEqual "Started"
+      refAfterRun mustEqual "Finished"
+    }
+
+    "cancel a job" in {
+      val JobId = 42
+
+      def jobStream(ref: SignallingRef[IO, String]): Stream[IO, Either[String, Int]] =
+        Stream.eval(ref.set("Started")).as(Right(1)) ++ await.as(Right(2)) ++ Stream.eval(ref.set("Working")).as(Right(3)) ++ await.as(Right(4)) ++ Stream.eval(ref.set("Finished")).as(Right(5))
```

Comment:
These `await`s aren't necessary for synchronization. They're meant to simulate work and give `fs2` a chance to cancel the job.

Reply:
Thanks for this clarification!